In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv("movies_metadata.csv")

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


ADULT = Binary Categorical Variable
COLLECTIONS = Harry potter,Kanchana ...
if collection = NAN means the movie's individuality 
collection column is dictionary datatype and genres is list datatype and so on
so it is a unstructured dataset

Budget column is continuous value but 0 represents it is missing value i.e no movies comes without budget

genres means action thriller, romantic , comedy
homepage means website for that movie
if a homepage exists for a movie which indicates that the movie is a high budget movie because small movies dont maintain a website....This is called as Derived Feature

overview is small description about the movie
revenue 0 represents missing value
runtime cant be more than the fixed maximum value .. if there then it may be an outlier

In [ ]:
data.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [ ]:
data.shape

(45466, 24)

In [ ]:
data.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

#1. Adult

Pandas taking Adult Values as Object ie... String but it should be boolean

In [ ]:
data.adult.values

array(['False', 'False', 'False', ..., 'False', 'False', 'False'],
      dtype=object)

In [ ]:
data.adult.unique() # finding Unique values in adult columns

array(['False', 'True', ' - Written by Ørnås',
       ' Rune Balot goes to a casino connected to the October corporation to try to wrap up her case once and for all.',
       ' Avalanche Sharks tells the story of a bikini contest that turns into a horrifying affair when it is hit by a shark avalanche.'],
      dtype=object)

In [ ]:
data.adult.isin(['True','False']) # in adult columns if machine finds True or False then it will print True if not then print False

0        True
1        True
2        True
3        True
4        True
         ... 
45461    True
45462    True
45463    True
45464    True
45465    True
Name: adult, Length: 45466, dtype: bool

In [ ]:
data.loc[~data.adult.isin(['True','False']),:] # opposite to previous step ie other than true or false everything will be printed


# not only adult columns some other columns are also wrong so drop that row


adult  \
19730                                 - Written by Ørnås   
29503   Rune Balot goes to a casino connected to the ...   
35587   Avalanche Sharks tells the story of a bikini ...   

      belongs_to_collection                            budget  \
19730              0.065736  /ff9qCepilowshEtG2GYWwzt2bs4.jpg   
29503              1.931659  /zV8bHuSL6WXoD6FWogP9j4x80bL.jpg   
35587              2.185485  /zaSf5OG7V8X8gqFvly88zDdRm46.jpg   

                                                  genres  \
19730  [{'name': 'Carousel Productions', 'id': 11176}...   
29503  [{'name': 'Aniplex', 'id': 2883}, {'name': 'Go...   
35587  [{'name': 'Odyssey Media', 'id': 17161}, {'nam...   

                                                homepage          id imdb_id  \
19730  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...  1997-08-20       0   
29503  [{'iso_3166_1': 'US', 'name': 'United States o...  2012-09-29       0   
35587           [{'iso_3166_1': 'CA', 'name': 'Canada'}]  2014-01-01       0   

      original_language                            original_title  overview  \
19730             104.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   
29503              68.0      [{'iso_639_1': 'ja', 'name': '日本語'}]  Released   
35587              82.0  [{'iso_639_1': 'en', 'name': 'English'}]  Released   

       ... release_date revenue runtime spoken_languages status  tagline  \
19730  ...            1     NaN     NaN              NaN    NaN      NaN   
29503  ...           12     NaN     NaN              NaN    NaN      NaN   
35587  ...           22     NaN     NaN              NaN    NaN      NaN   

       title video vote_average vote_count  
19730    NaN   NaN          NaN        NaN  
29503    NaN   NaN          NaN        NaN  
35587    NaN   NaN          NaN        NaN  

[3 rows x 24 columns]

In [ ]:
data = data.loc[data.adult.isin(['False','True']),:] # only choosing the rows where the adult column is either true or false
data.adult.unique()

array(['False', 'True'], dtype=object)

In [ ]:
data.adult.value_counts()

False    45454
True         9
Name: adult, dtype: int64

In [ ]:
data.adult = data.adult.map({'True':True,'False':False}) #convert from string to True/False

In [ ]:
#data.adult.dtype
data.adult.unique()

array([False,  True])

In [ ]:
data.adult.value_counts()

False    45454
True         9
Name: adult, dtype: int64

In [ ]:
data.shape

(45463, 24)

In [ ]:
data.adult.dtype

dtype('bool')

In [ ]:
data.adult.isnull().sum() # checking for NUll values

0

##2 . BELONGS TO COLLECTION

all movies need not to come under collections

In [ ]:
import numpy as np
data.belongs_to_collection.values[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [ ]:
data.belongs_to_collection.values[1]

nan

this column is read as string so change it into dictionary

nan means it doesnot belong to any collection of movies 

In [ ]:
import ast


In [ ]:
ast.literal_eval('[1,2,3]')

[1, 2, 3]

In [ ]:
data.belongs_to_collection.values[0]

"{'id': 10194, 'name': 'Toy Story Collection', 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}"

In [ ]:
ast.literal_eval(data.belongs_to_collection.values[0]) # ast will convert string into appropiate datatype... here we are checking for 1st column

{'id': 10194,
 'name': 'Toy Story Collection',
 'poster_path': '/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg',
 'backdrop_path': '/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'}

In [ ]:
ast.literal_eval(data.belongs_to_collection.values[0])['name']

'Toy Story Collection'

### whenever there exist a collection just take the name of the collection and if it is a nan value then it represents no collection

Having a dictionary in a column is a unstructured data 

In [ ]:
data.belongs_to_collection.values[1]

nan

In [ ]:
a=2
a==a

True

In [ ]:
a=np.nan
a==a

False

In [ ]:
collectionNames = [] #list to store collections names - list that is going to store the vleaned value
for i in data['belongs_to_collection'].values: #Iterating through the column
    if i == i: # check if it is nan 
      #not a nan value
      collectionNames.append(ast.literal_eval(i)['name'])
    else:
      # a nan value
      collectionNames.append("No Collection") #append -1

#collectionIds
data['collectionName'] = collectionNames
data = data.drop(columns = ['belongs_to_collection'])

In [ ]:
data

adult    budget                                             genres  \
0      False  30000000  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1      False  65000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2      False         0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3      False  16000000  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4      False         0                     [{'id': 35, 'name': 'Comedy'}]   
...      ...       ...                                                ...   
45461  False         0  [{'id': 18, 'name': 'Drama'}, {'id': 10751, 'n...   
45462  False         0                      [{'id': 18, 'name': 'Drama'}]   
45463  False         0  [{'id': 28, 'name': 'Action'}, {'id': 18, 'nam...   
45464  False         0                                                 []   
45465  False         0                                                 []   

                                   homepage      id    imdb_id  \
0      http://toystory.disney.com/toy-story     862  tt0114709   
1                                       NaN    8844  tt0113497   
2                                       NaN   15602  tt0113228   
3                                       NaN   31357  tt0114885   
4                                       NaN   11862  tt0113041   
...                                     ...     ...        ...   
45461  http://www.imdb.com/title/tt6209470/  439050  tt6209470   
45462                                   NaN  111109  tt2028550   
45463                                   NaN   67758  tt0303758   
45464                                   NaN  227506  tt0008536   
45465                                   NaN  461257  tt6980792   

      original_language               original_title  \
0                    en                    Toy Story   
1                    en                      Jumanji   
2                    en             Grumpier Old Men   
3                    en            Waiting to Exhale   
4                    en  Father of the Bride Part II   
...                 ...                          ...   
45461                fa                      رگ خواب   
45462                tl          Siglo ng Pagluluwal   
45463                en                     Betrayal   
45464                en          Satana likuyushchiy   
45465                en                     Queerama   

                                                overview popularity  ...  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
1      When siblings Judy and Peter discover an encha...  17.015539  ...   
2      A family wedding reignites the ancient feud be...    11.7129  ...   
3      Cheated on, mistreated and stepped on, the wom...   3.859495  ...   
4      Just when George Banks has recovered from his ...   8.387519  ...   
...                                                  ...        ...  ...   
45461        Rising and falling between a man and woman.   0.072051  ...   
45462  An artist struggles to finish his work while a...   0.178241  ...   
45463  When one of her hits goes wrong, a professiona...   0.903007  ...   
45464  In a small town live two brothers, one a minis...   0.003503  ...   
45465  50 years after decriminalisation of homosexual...   0.163015  ...   

           revenue runtime                                   spoken_languages  \
0      373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1      262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2              0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3       81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4       76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   
...            ...     ...                                                ...   
45461          0.0    90.0             [{'iso_639_1': 'fa', 'name': 'فارسی'}]   
45462          0.0   360.0                  [{'iso_639_

In [ ]:
data.collectionName.value_counts()

No Collection                    40972
The Bowery Boys                     29
Totò Collection                     27
James Bond Collection               26
Zatôichi: The Blind Swordsman       26
                                 ...  
Glass Tiger collection               1
Kathleen Madigan Collection          1
The Big Bottom Box                   1
Joséphine - Saga                     1
Red Lotus Collection                 1
Name: collectionName, Length: 1696, dtype: int64

#3.BUDGET

In [ ]:
data.budget.values

array(['30000000', '65000000', '0', ..., '0', '0', '0'], dtype=object)

In [ ]:
data.budget = data.budget.astype('int64') # explicitely convert string into integer

In [ ]:
data.budget.describe() #summary statistics

count    4.546300e+04
mean     4.224579e+06
std      1.742413e+07
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      3.800000e+08
Name: budget, dtype: float64

budget of a movie cannot be zero and if 0 exists then it is missing values

In [ ]:
sum(data.budget==0)          #how many movies contains budget = 0

36573

In [ ]:
data.shape[0]

45463

In [ ]:
sum(data.budget==0) / data.shape[0] * 100           # how many % of movies contains budget = 0

80.44563711149726

In a column there are 80 % of missing values we can either impute values from rest 20 % of data by using mean , median , mode or drop that column


but dropping that budget column is not correct method because without budget we cant find the revenue colllected by the movie

3 choices:
1 . drop budget column
2 . collect data from other dataset .. in this given case we can collect data by using imdb_id from imdb dataset
3 . delete 80 % of data and use rest 20 % data of budget column

as of now we dont have any problem statement for this dataset .. once data cleaning is done we will create a problem statement and do EDA

4 . Genres

In [ ]:
data.genres.values[0]

"[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]"

it ia unstructured data

In [ ]:
ast.literal_eval(data.genres.values[0])

[{'id': 16, 'name': 'Animation'},
 {'id': 35, 'name': 'Comedy'},
 {'id': 10751, 'name': 'Family'}]

In [ ]:
data.genres.unique()

array(["[{'id': 16, 'name': 'Animation'}, {'id': 35, 'name': 'Comedy'}, {'id': 10751, 'name': 'Family'}]",
       "[{'id': 12, 'name': 'Adventure'}, {'id': 14, 'name': 'Fantasy'}, {'id': 10751, 'name': 'Family'}]",
       "[{'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]",
       ...,
       "[{'id': 80, 'name': 'Crime'}, {'id': 35, 'name': 'Comedy'}, {'id': 28, 'name': 'Action'}, {'id': 10751, 'name': 'Family'}]",
       "[{'id': 28, 'name': 'Action'}, {'id': 9648, 'name': 'Mystery'}, {'id': 53, 'name': 'Thriller'}, {'id': 27, 'name': 'Horror'}]",
       "[{'id': 10751, 'name': 'Family'}, {'id': 16, 'name': 'Animation'}, {'id': 10749, 'name': 'Romance'}, {'id': 35, 'name': 'Comedy'}]"],
      dtype=object)

go through entire dataset
find unique genres
create unique number of columns

In [ ]:
#we are goig to looka t the data and find put all the unique genres that are there in my data
uniqueGenres = set()

for i in data.genres: # I am looping through all the rows
  genList = ast.literal_eval(i) # convert from string to list of dict
  for j in genList: # loop through the list of dict
    uniqueGenres.add(j['name']) # add the name of the genre to my uniqueGenres variabe


In [ ]:
print(uniqueGenres)
print(len(uniqueGenres))

{'History', 'Foreign', 'War', 'Action', 'Family', 'Documentary', 'Crime', 'Thriller', 'Animation', 'Music', 'Fantasy', 'Romance', 'Western', 'TV Movie', 'Drama', 'Comedy', 'Mystery', 'Science Fiction', 'Horror', 'Adventure'}
20


In [ ]:
rows = [] # for all the movies
genres = list(uniqueGenres) # all unique Genres

for i in data.genres: #looping through all the rows/movies
  genList = ast.literal_eval(i) #converting from string to list of dict for a particular movie
  movieGenres = [x['name'] for x in genList] # list of all the genres of the movie
  row = [] # store a list of 1s and 0s for one movie
  for j in genres: #loop through all unique genres
    row.append(int(j in movieGenres)) #checking if the genre is present in that movie - if the genre is present 1 else 0
  rows.append(row)

In [ ]:
rows[:10]

[[0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0],
 [0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1],
 [0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1],
 [0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]]

In [ ]:
genre = pd.DataFrame(rows,columns = genres)
genre

History  Foreign  War  Action  Family  Documentary  Crime  Thriller  \
0            0        0    0       0       1            0      0         0   
1            0        0    0       0       1            0      0         0   
2            0        0    0       0       0            0      0         0   
3            0        0    0       0       0            0      0         0   
4            0        0    0       0       0            0      0         0   
...        ...      ...  ...     ...     ...          ...    ...       ...   
45458        0        0    0       0       1            0      0         0   
45459        0        0    0       0       0            0      0         0   
45460        0        0    0       1       0            0      0         1   
45461        0        0    0       0       0            0      0         0   
45462        0        0    0       0       0            0      0         0   

       Animation  Music  Fantasy  Romance  Western  TV Movie  Drama  Comedy  \
0              1      0        0        0        0         0      0       1   
1              0      0        1        0        0         0      0       0   
2              0      0        0        1        0         0      0       1   
3              0      0        0        1        0         0      1       1   
4              0      0        0        0        0         0      0       1   
...          ...    ...      ...      ...      ...       ...    ...     ...   
45458          0      0        0        0        0         0      1       0   
45459          0      0        0        0        0         0      1       0   
45460          0      0        0        0        0         0      1       0   
45461          0      0        0        0        0         0      0       0   
45462          0      0        0        0        0         0      0       0   

       Mystery  Science Fiction  Horror  Adventure  
0            0                0       0          0  
1            0                0       0          1  
2            0                0       0          0  
3            0                0       0          0  
4            0                0       0          0  
...        ...              ...     ...        ...  
45458        0                0       0          0  
45459        0                0       0          0  
45460        0                0       0          0  
45461        0                0       0          0  
45462        0                0       0          0  

[45463 rows x 20 columns]

In [ ]:
genre.shape

(45463, 20)

In [ ]:
data.shape

(45463, 24)

In [ ]:
data = pd.concat([data,genre], axis = 1) # joing the two dataframes
data.shape, genre.shape

((45466, 44), (45463, 20))

In [ ]:
data = data.drop(columns = ['genres'])

In [ ]:
data

adult      budget                              homepage      id  \
0      False  30000000.0  http://toystory.disney.com/toy-story     862   
1      False  65000000.0                                   NaN    8844   
2      False         0.0                                   NaN   15602   
3      False  16000000.0                                   NaN   31357   
4      False         0.0                                   NaN   11862   
...      ...         ...                                   ...     ...   
45461  False         0.0  http://www.imdb.com/title/tt6209470/  439050   
45462  False         0.0                                   NaN  111109   
45463  False         0.0                                   NaN   67758   
45464  False         0.0                                   NaN  227506   
45465  False         0.0                                   NaN  461257   

         imdb_id original_language               original_title  \
0      tt0114709                en                    Toy Story   
1      tt0113497                en                      Jumanji   
2      tt0113228                en             Grumpier Old Men   
3      tt0114885                en            Waiting to Exhale   
4      tt0113041                en  Father of the Bride Part II   
...          ...               ...                          ...   
45461  tt6209470                fa                      رگ خواب   
45462  tt2028550                tl          Siglo ng Pagluluwal   
45463  tt0303758                en                     Betrayal   
45464  tt0008536                en          Satana likuyushchiy   
45465  tt6980792                en                     Queerama   

                                                overview popularity  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943   
1      When siblings Judy and Peter discover an encha...  17.015539   
2      A family wedding reignites the ancient feud be...    11.7129   
3      Cheated on, mistreated and stepped on, the wom...   3.859495   
4      Just when George Banks has recovered from his ...   8.387519   
...                                                  ...        ...   
45461        Rising and falling between a man and woman.   0.072051   
45462  An artist struggles to finish his work while a...   0.178241   
45463  When one of her hits goes wrong, a professiona...   0.903007   
45464  In a small town live two brothers, one a minis...   0.003503   
45465  50 years after decriminalisation of homosexual...   0.163015   

                            poster_path  ... Fantasy Romance Western  \
0      /rhIRbceoE9lR4veEXuwCC2wARtG.jpg  ...     0.0     0.0     0.0   
1      /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg  ...     1.0     0.0     0.0   
2      /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg  ...     0.0     1.0     0.0   
3      /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg  ...     0.0     1.0     0.0   
4      /e64sOI48hQXyru7naBFyssKFxVd.jpg  ...     0.0     0.0     0.0   
...                                 ...  ...     ...     ...     ...   
45461  /jldsYflnId4tTWPx8es3uzsB1I8.jpg  ...     0.0     0.0     0.0   
45462  /xZkmxsNmYXJbKVsTRLLx3pqGHx7.jpg  ...     0.0     0.0     0.0   
45463  /d5bX92nDsISNhu3ZT69uHwmfCGw.jpg  ...     NaN     NaN     NaN   
45464  /aorBPO7ak8e8iJKT5OcqYxU3jlK.jpg  ...     NaN     NaN     NaN   
45465  /s5UkZt6NTsrS7ZF0Rh8nzupRlIU.jpg  ...     NaN     NaN     NaN   

       TV Movie  Drama Comedy Mystery Science Fiction Horror Adventure  
0           0.0    0.0    1.0     0.0             0.0    0.0       0.0  
1           0.0    0.0    0.0     0.0             0.0    0.0       1.0  
2           0.0    0.0    1.0     0.0             0.0    0.0       0.0  
3           0.0    1.0    1.0     0.0             0.0    0.0       0.0  
4           0.0    0.0    1.0     0.0             0.0    0.0       0.0  
...         ...    ...    ...     ...             ...    ...       ...  
45461       0.0    0.0    0.0     0.0             0.0    0.0       0.0  
45462      

#5 . homepage 

In [ ]:
data.homepage.values[0]

'http://toystory.disney.com/toy-story'

In [ ]:
data.homepage.values[1]

nan

rather than websites whether the movie has websites or not is going to add values to the data

In [ ]:
data.homepage.isnull() # print true if it is nan value

0        False
1         True
2         True
3         True
4         True
         ...  
45461    False
45462     True
45463     True
45464     True
45465     True
Name: homepage, Length: 45466, dtype: bool

In [ ]:
~data.homepage.isnull() 

0         True
1        False
2        False
3        False
4        False
         ...  
45461     True
45462    False
45463    False
45464    False
45465    False
Name: homepage, Length: 45466, dtype: bool

In [ ]:
data.homepage.isnull().sum()*100/data.shape[0]

82.89051159107905

In [ ]:
data['homePagePresent'] = (~data.homepage.isnull()) #false- if null in home page else true

In [ ]:
data = data.drop(columns = ['homepage'])

In [ ]:
data.homePagePresent = data.homePagePresent.map({True:1,False:0})

In [ ]:
data

adult      budget      id    imdb_id original_language  \
0      False  30000000.0     862  tt0114709                en   
1      False  65000000.0    8844  tt0113497                en   
2      False         0.0   15602  tt0113228                en   
3      False  16000000.0   31357  tt0114885                en   
4      False         0.0   11862  tt0113041                en   
...      ...         ...     ...        ...               ...   
45461  False         0.0  439050  tt6209470                fa   
45462  False         0.0  111109  tt2028550                tl   
45463  False         0.0   67758  tt0303758                en   
45464  False         0.0  227506  tt0008536                en   
45465  False         0.0  461257  tt6980792                en   

                    original_title  \
0                        Toy Story   
1                          Jumanji   
2                 Grumpier Old Men   
3                Waiting to Exhale   
4      Father of the Bride Part II   
...                            ...   
45461                      رگ خواب   
45462          Siglo ng Pagluluwal   
45463                     Betrayal   
45464          Satana likuyushchiy   
45465                     Queerama   

                                                overview popularity  \
0      Led by Woody, Andy's toys live happily in his ...  21.946943   
1      When siblings Judy and Peter discover an encha...  17.015539   
2      A family wedding reignites the ancient feud be...    11.7129   
3      Cheated on, mistreated and stepped on, the wom...   3.859495   
4      Just when George Banks has recovered from his ...   8.387519   
...                                                  ...        ...   
45461        Rising and falling between a man and woman.   0.072051   
45462  An artist struggles to finish his work while a...   0.178241   
45463  When one of her hits goes wrong, a professiona...   0.903007   
45464  In a small town live two brothers, one a minis...   0.003503   
45465  50 years after decriminalisation of homosexual...   0.163015   

                            poster_path  \
0      /rhIRbceoE9lR4veEXuwCC2wARtG.jpg   
1      /vzmL6fP7aPKNKPRTFnZmiUfciyV.jpg   
2      /6ksm1sjKMFLbO7UY2i6G1ju9SML.jpg   
3      /16XOMpEaLWkrcPqSQqhTmeJuqQl.jpg   
4      /e64sOI48hQXyru7naBFyssKFxVd.jpg   
...                                 ...   
45461  /jldsYflnId4tTWPx8es3uzsB1I8.jpg   
45462  /xZkmxsNmYXJbKVsTRLLx3pqGHx7.jpg   
45463  /d5bX92nDsISNhu3ZT69uHwmfCGw.jpg   
45464  /aorBPO7ak8e8iJKT5OcqYxU3jlK.jpg   
45465  /s5UkZt6NTsrS7ZF0Rh8nzupRlIU.jpg   

                                    production_companies  ... Romance Western  \
0         [{'name': 'Pixar Animation Studios', 'id': 3}]  ...     0.0     0.0   
1      [{'name': 'TriStar Pictures', 'id': 559}, {'na...  ...     0.0     0.0   
2      [{'name': 'Warner Bros.', 'id': 6194}, {'name'...  ...     1.0     0.0   
3      [{'name': 'Twentieth Century Fox Film Corporat...  ...     1.0     0.0   
4      [{'name': 'Sandollar Productions', 'id': 5842}...  ...     0.0     0.0   
...                                                  ...  ...     ...     ...   
45461                                                 []  ...     0.0     0.0   
45462             [{'name': 'Sine Olivia', 'id': 19653}]  ...     0.0     0.0   
45463  [{'name': 'American World Pictures', 'id': 6165}]  ...     NaN     NaN   
45464               [{'name': 'Yermoliev', 'id': 88753}]  ...     NaN     NaN   
45465                                                 []  ...     NaN     NaN   

       TV Movie  Drama Comedy Mystery Science Fiction Horror Adventure  \
0           0.0    0.0    1.0     0.0             0.0    0.0       0.0   
1           0.0    0.0    0.0     0.0             0.0    0.0       1.0   
2           0.0    0.0    1.0     0.0             0.0    0.0       0.0   
3           0.0    1.0    1.0     0.0             0.0    0.0       0.0   
4           0.0    0.0    1.0     0.0             0.0   

#6 . ID

1 . Duplicates

2 . Missing value

3 . Int data type

In [ ]:
data.id.unique()

array(['862', '8844', '15602', ..., '67758', '227506', '461257'],
      dtype=object)

In [ ]:
len(data.id.unique())

45434

In [ ]:
data.shape[0] - len(data.id.unique())   #difference between movies and unique ids

32

In [ ]:
data = data.drop_duplicates(subset = 'id') # delete the rows which has id column to be the same
data.shape

(45434, 43)

In [ ]:
data.shape[0] - len(data.id.unique())  

0

In [ ]:
data.id.values[0]

'862'

In [ ]:
#data.id = data.id.astype('int64')

In [ ]:
data = data.dropna(subset = ['id'] )

In [ ]:
data.id = data.id.astype('int64')

In [ ]:
data.id

0           862
1          8844
2         15602
3         31357
4         11862
          ...  
45461    439050
45462    111109
45463     67758
45464    227506
45465    461257
Name: id, Length: 45433, dtype: int64

There are 3 task to be done here
 1 . delete duplicates
 2 . delete missing values
 3 . change from string to int

#7 . imdb_id

repeat same thing for imdb_id columns also

In [ ]:
data.imdb_id.values[0]

'tt0114709'

In [ ]:
data.imdb_id.values[0]
(data.shape[0] - len(data.imdb_id.unique()))
data = data.drop_duplicates(subset = 'imdb_id')

#8 . Original languages

In [ ]:
data.original_language.values[0]

'en'

In [ ]:
len(data.original_language.unique())

90

In [ ]:
data.original_language.value_counts()

en    32239
fr     2435
it     1529
ja     1346
de     1079
      ...  
ay        1
rw        1
cy        1
tg        1
si        1
Name: original_language, Length: 89, dtype: int64

In [ ]:
data.original_language.isnull().sum()

11

fill null values with mode becoz it is cateogorical data

In [ ]:
data['original_language'] = data.original_language.fillna('mode')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
data.columns

Index(['adult', 'budget', 'id', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'poster_path',
       'production_companies', 'production_countries', 'release_date',
       'revenue', 'runtime', 'spoken_languages', 'status', 'tagline', 'title',
       'video', 'vote_average', 'vote_count', 'collectionName', 'History',
       'Foreign', 'War', 'Action', 'Family', 'Documentary', 'Crime',
       'Thriller', 'Animation', 'Music', 'Fantasy', 'Romance', 'Western',
       'TV Movie', 'Drama', 'Comedy', 'Mystery', 'Science Fiction', 'Horror',
       'Adventure', 'homePagePresent'],
      dtype='object')

#9 . title

In [ ]:
data.title.isnull().sum(), len(data.title.unique())*100/data.shape[0] # (null value . % of unique value )

(3, 93.05986745051412)

In [ ]:
data.original_title.isnull().sum(), len(data.original_title.unique())*100/data.shape[0]

(0, 95.464253473369)

In [ ]:
data = data.drop(columns = ['title'])

In [ ]:
data.original_title.values[0]

'Toy Story'

In [ ]:
'bala'=='bala '

False

1) lower case upper case
2) remove unwanted spaces (data science, data  science)
3) checking spelling mistakes -> for places where there are a lot of chracters (textBlob)
4) handling special characters

In [ ]:
data.original_title = data['original_title'].str.lower() # brinf the text data to lowercase
data.original_title = data['original_title'].str.strip() # delete unwanted whtespaces

In [ ]:
data.original_title

0                          toy story
1                            jumanji
2                   grumpier old men
3                  waiting to exhale
4        father of the bride part ii
                    ...             
45461                        رگ خواب
45462            siglo ng pagluluwal
45463                       betrayal
45464            satana likuyushchiy
45465                       queerama
Name: original_title, Length: 45417, dtype: object

#10 . overview


it is text data so we should do NLP and so we are doing some basic cleaning here

In [ ]:
data.overview.values[0]

"Led by Woody, Andy's toys live happily in his room until Andy's birthday brings Buzz Lightyear onto the scene. Afraid of losing his place in Andy's heart, Woody plots against Buzz. But when circumstances separate Buzz and Woody from their owner, the duo eventually learns to put aside their differences."

In [ ]:
data.overview = data.overview.str.lower()
data.overview = data.overview.str.strip()
data.overview

0        led by woody, andy's toys live happily in his ...
1        when siblings judy and peter discover an encha...
2        a family wedding reignites the ancient feud be...
3        cheated on, mistreated and stepped on, the wom...
4        just when george banks has recovered from his ...
                               ...                        
45461          rising and falling between a man and woman.
45462    an artist struggles to finish his work while a...
45463    when one of her hits goes wrong, a professiona...
45464    in a small town live two brothers, one a minis...
45465    50 years after decriminalisation of homosexual...
Name: overview, Length: 45417, dtype: object

In [ ]:
data.overview.isnull().sum()

952

In [ ]:
data.loc[data.overview.isnull(),'overview'] = 'no overview available'

# 11 . Popularity

Popularity is calculated by movie lens website through some formula like no of search , views

In [ ]:
data.popularity.values[5]

17.924927

In [ ]:
data.popularity.dtype

dtype('O')

In [ ]:
data.popularity = data.popularity.astype('float64') ##floating point value
data.popularity.describe()

count    45414.000000
mean         2.921932
std          6.007588
min          0.000000
25%          0.386107
50%          1.127685
75%          3.679143
max        547.488298
Name: popularity, dtype: float64

possibility of outlier

In [ ]:
#IQR
q3 = data.popularity.quantile(0.75)
q1 = data.popularity.quantile(0.25)
iqr = q3 - q1 

In [ ]:
#thresholds
q3 + (1.5 * iqr)

8.618696

In [ ]:
q1 - (1.5*iqr)

-4.553446

In [ ]:
##outlier detection - 1) domain knowledge 2) u need to handle them
data.popularity = data.popularity.clip(-4.55,8.62) #.clip(lt,ut)
#data = data.loc[(data.popularity > 0) & (data.popularity <= 100),:]

In [ ]:
data.popularity.describe()

count    45414.000000
mean         2.459170
std          2.831027
min          0.000000
25%          0.386107
50%          1.127685
75%          3.679143
max          8.620000
Name: popularity, dtype: float64

In [ ]:
data.popularity.isnull().sum()

3

In [ ]:
data = data.dropna(subset = ['popularity']) #find all the null values in my popularity column and delete them 

In [ ]:
data.popularity.isnull().sum()

0

#12 . Vote count 

In [ ]:
data.vote_count.values[0]

5415.0

In [ ]:
data.vote_count.dtype

dtype('float64')

In [ ]:
data.vote_count = data.vote_count.astype(int)

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
sum(data.vote_count.isnull())

0

In [ ]:
data.vote_count.describe()

count    45414.000000
mean       109.971837
std        491.549122
min          0.000000
25%          3.000000
50%         10.000000
75%         34.000000
max      14075.000000
Name: vote_count, dtype: float64

no people voted for a movie and 14075 people voting for a movie ... both are possible

no outlier correction is needed

In [ ]:
data.vote_count.isna().sum()    # checking for null values

0

# 13 . Vote average

In [ ]:
data.vote_average.values[0]

7.7

In [ ]:
data.vote_average.dtype

dtype('float64')

In [ ]:
data.vote_average.isnull().sum()

0

In [ ]:
data.vote_average.describe()

count    45414.000000
mean         5.618864
std          1.923210
min          0.000000
25%          5.000000
50%          6.000000
75%          6.800000
max         10.000000
Name: vote_average, dtype: float64

In [ ]:
# data['score'] = (data.vote_average * data.vote_count)    derived feature

#14 . Poster path

In [ ]:
#data.poster_path.values[0] #null value, unique values, file exists in the path or not

In [ ]:
#data.poster_path.isnull().sum()

This code will check whether that particular file present in the colab or not 

In [ ]:
import os
os.path.isfile('/content/movies_metadata.csv')

True

In [ ]:
os.path.isfile('/content/bala.csv')

False

In [ ]:
'''import os
#data.poster_path.map(lambda x  : os.path.isfile(str(x)))
os.path.isfile('/content/movies_metadata.csv')
#os.mkdir("/content/koushik")'''

'import os\n#data.poster_path.map(lambda x  : os.path.isfile(str(x)))\nos.path.isfile(\'/content/movies_metadata.csv\')\n#os.mkdir("/content/koushik")'

In [ ]:
data = data.drop(columns = ['poster_path'])

# 15 . Production Companies


In [ ]:
data.production_companies.values[0]

"[{'name': 'Pixar Animation Studios', 'id': 3}]"

In [ ]:
data.production_companies.isnull().sum()

0

In [ ]:
data.production_companies.values[0]

"[{'name': 'Pixar Animation Studios', 'id': 3}]"

In [ ]:
#Code for extracting Producer 1
import ast
index = 0
prod1 = [] #empty list which is going to store producer 1's name for each movie

for prod in data.production_companies.values: # looping through all rows/movies
  prodList = ast.literal_eval(prod) #string -> list of dicts #for each movie, prodList will contain the set of all the producers
  if len(prodList) > index:  # if they have atleast 1 producer
    prod1.append(prodList[index]['name']) #note the name of producer
  else:
    prod1.append('No Producer') #put them as no producer

data['Producer1'] = prod1

In [ ]:
# Code for extracting producer 2
import ast
index = 1
prod2 = []

for prod in data.production_companies.values:
  prodList = ast.literal_eval(prod)
  if len(prodList) > index:
    prod2.append(prodList[index]['name'])
  else:
    prod2.append('No Producer')

data['Producer2'] = prod2

In [ ]:
data

adult      budget      id    imdb_id original_language  \
0      False  30000000.0     862  tt0114709                en   
1      False  65000000.0    8844  tt0113497                en   
2      False         0.0   15602  tt0113228                en   
3      False  16000000.0   31357  tt0114885                en   
4      False         0.0   11862  tt0113041                en   
...      ...         ...     ...        ...               ...   
45461  False         0.0  439050  tt6209470                fa   
45462  False         0.0  111109  tt2028550                tl   
45463  False         0.0   67758  tt0303758                en   
45464  False         0.0  227506  tt0008536                en   
45465  False         0.0  461257  tt6980792                en   

                    original_title  \
0                        toy story   
1                          jumanji   
2                 grumpier old men   
3                waiting to exhale   
4      father of the bride part ii   
...                            ...   
45461                      رگ خواب   
45462          siglo ng pagluluwal   
45463                     betrayal   
45464          satana likuyushchiy   
45465                     queerama   

                                                overview  popularity  \
0      led by woody, andy's toys live happily in his ...    8.620000   
1      when siblings judy and peter discover an encha...    8.620000   
2      a family wedding reignites the ancient feud be...    8.620000   
3      cheated on, mistreated and stepped on, the wom...    3.859495   
4      just when george banks has recovered from his ...    8.387519   
...                                                  ...         ...   
45461        rising and falling between a man and woman.    0.072051   
45462  an artist struggles to finish his work while a...    0.178241   
45463  when one of her hits goes wrong, a professiona...    0.903007   
45464  in a small town live two brothers, one a minis...    0.003503   
45465  50 years after decriminalisation of homosexual...    0.163015   

                                    production_companies  \
0         [{'name': 'Pixar Animation Studios', 'id': 3}]   
1      [{'name': 'TriStar Pictures', 'id': 559}, {'na...   
2      [{'name': 'Warner Bros.', 'id': 6194}, {'name'...   
3      [{'name': 'Twentieth Century Fox Film Corporat...   
4      [{'name': 'Sandollar Productions', 'id': 5842}...   
...                                                  ...   
45461                                                 []   
45462             [{'name': 'Sine Olivia', 'id': 19653}]   
45463  [{'name': 'American World Pictures', 'id': 6165}]   
45464               [{'name': 'Yermoliev', 'id': 88753}]   
45465                                                 []   

                                    production_countries  ... TV Movie  Drama  \
0      [{'iso_3166_1': 'US', 'name': 'United States o...  ...      0.0    0.0   
1      [{'iso_3166_1': 'US', 'name': 'United States o...  ...      0.0    0.0   
2      [{'iso_3166_1': 'US', 'name': 'United States o...  ...      0.0    0.0   
3      [{'iso_3166_1': 'US', 'name': 'United States o...  ...      0.0    1.0   
4      [{'iso_3166_1': 'US', 'name': 'United States o...  ...      0.0    0.0   
...                                                  ...  ...      ...    ...   
45461             [{'iso_3166_1': 'IR', 'name': 'Iran'}]  ...      0.0    0.0   
45462      [{'iso_3166_1': 'PH', 'name': 'Philippines'}]  ...      0.0    0.0   
45463  [{'iso_3166_1': 'US', 'name': 'United States o...  ...      NaN    NaN   
45464           [{'iso_3166_1': 'RU', 'name': 'Russia'}]  ...      NaN    NaN   
45465   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]  ...      NaN    NaN   

       Comedy Mystery Science Fiction Horror Adventure  homePagePresent  \
0         1.0     0.0             0.0    0.0       0.0                1   
1         0.0     0.0             0.0    0.0       1.0      

In [ ]:
data = data.drop(columns = ['production_companies'])

In [ ]:
data

adult      budget      id    imdb_id original_language  \
0      False  30000000.0     862  tt0114709                en   
1      False  65000000.0    8844  tt0113497                en   
2      False         0.0   15602  tt0113228                en   
3      False  16000000.0   31357  tt0114885                en   
4      False         0.0   11862  tt0113041                en   
...      ...         ...     ...        ...               ...   
45461  False         0.0  439050  tt6209470                fa   
45462  False         0.0  111109  tt2028550                tl   
45463  False         0.0   67758  tt0303758                en   
45464  False         0.0  227506  tt0008536                en   
45465  False         0.0  461257  tt6980792                en   

                    original_title  \
0                        toy story   
1                          jumanji   
2                 grumpier old men   
3                waiting to exhale   
4      father of the bride part ii   
...                            ...   
45461                      رگ خواب   
45462          siglo ng pagluluwal   
45463                     betrayal   
45464          satana likuyushchiy   
45465                     queerama   

                                                overview  popularity  \
0      led by woody, andy's toys live happily in his ...    8.620000   
1      when siblings judy and peter discover an encha...    8.620000   
2      a family wedding reignites the ancient feud be...    8.620000   
3      cheated on, mistreated and stepped on, the wom...    3.859495   
4      just when george banks has recovered from his ...    8.387519   
...                                                  ...         ...   
45461        rising and falling between a man and woman.    0.072051   
45462  an artist struggles to finish his work while a...    0.178241   
45463  when one of her hits goes wrong, a professiona...    0.903007   
45464  in a small town live two brothers, one a minis...    0.003503   
45465  50 years after decriminalisation of homosexual...    0.163015   

                                    production_countries release_date  ...  \
0      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30  ...   
1      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15  ...   
2      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22  ...   
3      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22  ...   
4      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10  ...   
...                                                  ...          ...  ...   
45461             [{'iso_3166_1': 'IR', 'name': 'Iran'}]          NaN  ...   
45462      [{'iso_3166_1': 'PH', 'name': 'Philippines'}]   2011-11-17  ...   
45463  [{'iso_3166_1': 'US', 'name': 'United States o...   2003-08-01  ...   
45464           [{'iso_3166_1': 'RU', 'name': 'Russia'}]   1917-10-21  ...   
45465   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]   2017-06-09  ...   

       TV Movie  Drama Comedy Mystery Science Fiction Horror  Adventure  \
0           0.0    0.0    1.0     0.0             0.0    0.0        0.0   
1           0.0    0.0    0.0     0.0             0.0    0.0        1.0   
2           0.0    0.0    1.0     0.0             0.0    0.0        0.0   
3           0.0    1.0    1.0     0.0             0.0    0.0        0.0   
4           0.0    0.0    1.0     0.0             0.0    0.0        0.0   
...         ...    ...    ...     ...             ...    ...        ...   
45461       0.0    0.0    0.0     0.0             0.0    0.0        0.0   
45462       0.0    0.0    0.0     0.0             0.0    0.0        0.0   
45463       NaN    NaN    NaN     NaN             NaN    NaN        NaN   
45464       NaN    NaN    NaN     NaN             NaN    NaN        NaN   
45465       NaN    NaN    NaN     NaN             NaN    NaN        NaN   

       homePagePresent                               Producer1  \


# 16 . Release Dates

In [ ]:
data.release_date.values[0] #  datatype

'1995-10-30'

In [ ]:
data.release_date.isnull().sum()

81

In [ ]:
data = data.dropna(subset = ['release_date'])

In [ ]:
data.release_date = pd.to_datetime(data.release_date, format='%Y-%m-%d')

/usr/local/lib/python3.7/dist-packages/pandas/core/generic.py:5516: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [ ]:
data.release_date # ,data.release_dat.dt.month,dt.weekday,dt.year 

0       1995-10-30
1       1995-12-15
2       1995-12-22
3       1995-12-22
4       1995-02-10
           ...    
45460   1991-05-13
45462   2011-11-17
45463   2003-08-01
45464   1917-10-21
45465   2017-06-09
Name: release_date, Length: 45333, dtype: datetime64[ns]

#17 . Revenues

In [ ]:
data.revenue.values[0]

373554033.0

In [ ]:
data.revenue.dtype

dtype('float64')

In [ ]:
data.revenue = data.revenue.astype('int64')

In [ ]:
data.revenue.describe()

count    4.533300e+04
mean     1.123688e+07
std      6.441885e+07
min      0.000000e+00
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      2.787965e+09
Name: revenue, dtype: float64

In [ ]:
sum(data.revenue == 0)*100/data.shape[0] # % of data having missing value

83.68296825711954

#18 . Runtime

In [ ]:
data.runtime.values[0]

81.0

In [ ]:
data.runtime.describe()

count    45089.000000
mean        94.184014
std         38.342419
min          0.000000
25%         85.000000
50%         95.000000
75%        107.000000
max       1256.000000
Name: runtime, dtype: float64

In [ ]:
sum(data.runtime == 0)

1532

In [ ]:
data.loc[data.runtime == 0,'runtime']  = 95   # imputation through median ( i.e ) 50 % quantile

/usr/local/lib/python3.7/dist-packages/pandas/core/indexing.py:1817: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [ ]:
sum(data.runtime > 240) # movies which are more than 4 hrs runtime mostly they are series , documentary

194

In [ ]:
data.loc[data.runtime > 240, :]

adult      budget      id    imdb_id original_language  \
1260   False  23000000.0   10341  tt0101458                de   
1360   False         0.0   10549  tt0116477                en   
3409   False         0.0   61560  tt0108167                fr   
3922   False         0.0   10655  tt0107007                en   
4146   False  31115000.0    8095  tt0056937                en   
...      ...         ...     ...        ...               ...   
44525  False         0.0  176167  tt0084476                de   
44721  False         0.0  458298  tt6792200                en   
44758  False         0.0  128733  tt0080230                en   
45424  False         0.0   42652  tt0096356                ru   
45462  False         0.0  111109  tt2028550                tl   

              original_title  \
1260   bis ans ende der welt   
1360                  hamlet   
3409    smoking / no smoking   
3922              gettysburg   
4146               cleopatra   
...                      ...   
44525               parsifal   
44721            the keepers   
44758              hollywood   
45424         uznik zamka if   
45462    siglo ng pagluluwal   

                                                overview  popularity  \
1260   until the end of the world is an odyssey for t...    2.146520   
1360   hamlet, prince of denmark, returns home to fin...    8.620000   
3409   "smoking" and "no smoking" are two segments of...    0.501632   
3922   summer 1863. the confederacy pushes north into...    6.588347   
4146   historical epic. the triumphs and tragedy of t...    8.591238   
...                                                  ...         ...   
44525  hans-jürgen syberberg's parsifal film is a psy...    0.165026   
44721  a seven-part docuseries about the unsolved mur...    1.576170   
44758  the definitive documentary about the american ...    0.055713   
45424  a russian version of the classic alexandre dum...    0.243401   
45462  an artist struggles to finish his work while a...    0.178241   

                                    production_countries release_date  ...  \
1260   [{'iso_3166_1': 'AU', 'name': 'Australia'}, {'...   1991-09-12  ...   
1360   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   1996-12-25  ...   
3409            [{'iso_3166_1': 'FR', 'name': 'France'}]   1993-12-15  ...   
3922   [{'iso_3166_1': 'US', 'name': 'United States o...   1993-10-08  ...   
4146   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'...   1963-06-12  ...   
...                                                  ...          ...  ...   
44525          [{'iso_3166_1': 'DE', 'name': 'Germany'}]   1982-04-30  ...   
44721  [{'iso_3166_1': 'US', 'name': 'United States o...   2017-05-19  ...   
44758                                                 []   1980-02-09  ...   
45424  [{'iso_3166_1': 'UA', 'name': 'Ukraine'}, {'is...   1988-01-01  ...   
45462      [{'iso_3166_1': 'PH', 'name': 'Philippines'}]   2011-11-17  ...   

       TV Movie  Drama Comedy Mystery Science Fiction Horror  Adventure  \
1260        0.0    1.0    0.0     0.0             1.0    0.0        0.0   
1360        0.0    1.0    0.0     0.0             0.0    0.0        0.0   
3409        0.0    1.0    1.0     0.0             0.0    0.0        0.0   
3922        0.0    1.0    0.0     0.0             0.0    0.0        0.0   
4146        0.0    1.0    0.0     0.0             0.0    0.0        0.0   
...         ...    ...    ...     ...             ...    ...        ...   
44525       0.0    1.0    1.0     0.0             0.0    0.0        0.0   
44721       0.0    0.0    0.0     0.0             0.0    0.0        0.0   
44758       0.0    0.0    0.0     0.0             0.0    1.0        0.0   
45424       0.0    1.0    0.0     1.0             0.0    1.0        0.0   
45462       0.0    0.0    0.0     0.0             0.0    0.0        0.0   

       homePagePresent                               Producer1  \
1260                 0               Village Roadshow Pictures   
1360  

In [ ]:
data = data.loc[data.runtime <= 240,:]

In [ ]:
data.runtime.isnull().sum()

0

#19 . Spoken Languages

In [ ]:
data.spoken_languages.values[0]

"[{'iso_639_1': 'en', 'name': 'English'}]"

In [ ]:
import ast
totalLanguages = []

for prod in data.spoken_languages.values: # looping through all the rows
  prodList = ast.literal_eval(prod) #converting from string to list of dictionary 
  totalLanguages.append(len(prodList)) #how many languages spoken?

data['TotalLanguages'] = totalLanguages

In [ ]:
data = data.drop(columns = ['spoken_languages'])

In [ ]:
data

adult      budget      id    imdb_id original_language  \
0      False  30000000.0     862  tt0114709                en   
1      False  65000000.0    8844  tt0113497                en   
2      False         0.0   15602  tt0113228                en   
3      False  16000000.0   31357  tt0114885                en   
4      False         0.0   11862  tt0113041                en   
...      ...         ...     ...        ...               ...   
45459  False         0.0  222848  tt0112613                en   
45460  False         0.0   30840  tt0102797                en   
45463  False         0.0   67758  tt0303758                en   
45464  False         0.0  227506  tt0008536                en   
45465  False         0.0  461257  tt6980792                en   

                    original_title  \
0                        toy story   
1                          jumanji   
2                 grumpier old men   
3                waiting to exhale   
4      father of the bride part ii   
...                            ...   
45459              caged heat 3000   
45460                   robin hood   
45463                     betrayal   
45464          satana likuyushchiy   
45465                     queerama   

                                                overview  popularity  \
0      led by woody, andy's toys live happily in his ...    8.620000   
1      when siblings judy and peter discover an encha...    8.620000   
2      a family wedding reignites the ancient feud be...    8.620000   
3      cheated on, mistreated and stepped on, the wom...    3.859495   
4      just when george banks has recovered from his ...    8.387519   
...                                                  ...         ...   
45459  it's the year 3000 ad. the world's most danger...    0.661558   
45460  yet another version of the classic epic, with ...    5.683753   
45463  when one of her hits goes wrong, a professiona...    0.903007   
45464  in a small town live two brothers, one a minis...    0.003503   
45465  50 years after decriminalisation of homosexual...    0.163015   

                                    production_countries release_date  ...  \
0      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30  ...   
1      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15  ...   
2      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22  ...   
3      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22  ...   
4      [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10  ...   
...                                                  ...          ...  ...   
45459  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-01-01  ...   
45460  [{'iso_3166_1': 'CA', 'name': 'Canada'}, {'iso...   1991-05-13  ...   
45463  [{'iso_3166_1': 'US', 'name': 'United States o...   2003-08-01  ...   
45464           [{'iso_3166_1': 'RU', 'name': 'Russia'}]   1917-10-21  ...   
45465   [{'iso_3166_1': 'GB', 'name': 'United Kingdom'}]   2017-06-09  ...   

       Drama  Comedy Mystery Science Fiction Horror  Adventure  \
0        0.0     1.0     0.0             0.0    0.0        0.0   
1        0.0     0.0     0.0             0.0    0.0        1.0   
2        0.0     1.0     0.0             0.0    0.0        0.0   
3        1.0     1.0     0.0             0.0    0.0        0.0   
4        0.0     1.0     0.0             0.0    0.0        0.0   
...      ...     ...     ...             ...    ...        ...   
45459    1.0     0.0     0.0             0.0    0.0        0.0   
45460    1.0     0.0     0.0             0.0    0.0        0.0   
45463    NaN     NaN     NaN             NaN    NaN        NaN   
45464    NaN     NaN     NaN             NaN    NaN        NaN   
45465    NaN     NaN     NaN             NaN    NaN        NaN   

       homePagePresent                               Producer1  \
0                    1                 Pixar Animation Studios   
1                    0                    

#20 .Status

In [ ]:
data.status.values[0]

'Released'

In [ ]:
data.status.value_counts()

Released           44464
Rumored              227
Post Production       97
In Production         17
Planned               13
Canceled               1
Name: status, dtype: int64

In [ ]:
data.status = data.status.fillna('Released')

In [ ]:
data.status.isnull().sum()

0

In [ ]:
data.columns

Index(['adult', 'budget', 'id', 'imdb_id', 'original_language',
       'original_title', 'overview', 'popularity', 'production_countries',
       'release_date', 'revenue', 'runtime', 'status', 'tagline', 'video',
       'vote_average', 'vote_count', 'collectionName', 'History', 'Foreign',
       'War', 'Action', 'Family', 'Documentary', 'Crime', 'Thriller',
       'Animation', 'Music', 'Fantasy', 'Romance', 'Western', 'TV Movie',
       'Drama', 'Comedy', 'Mystery', 'Science Fiction', 'Horror', 'Adventure',
       'homePagePresent', 'Producer1', 'Producer2', 'TotalLanguages'],
      dtype='object')

#21 . Others

In [ ]:
data = data.drop(columns = ['tagline','video'])

In [ ]:
data = data.drop(columns = ['production_countries'])

#22.Final

In [ ]:
data = data.dropna()

In [ ]:
data.shape

(44891, 39)

In [ ]:
dataBudgetRevenue  = data.loc[(data.budget != 0) & (data.revenue != 0),:]
dataBudgetRevenue.shape

(5371, 39)

# profit loss calculation

In [ ]:
dataBudgetRevenue['profit'] = (dataBudgetRevenue.revenue - dataBudgetRevenue.budget)/dataBudgetRevenue.budget
dataBudgetRevenue['profit'] = dataBudgetRevenue['profit'].clip(-1,10) # -1 --> 100 % loss 10 --> 100 % prifit

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
dataBudgetRevenue.to_csv('finalCleanedData.csv')

1. Profit = (Revenue - Budget)/Budget
2. Do the movies that have high vote_avergae make lot of profits ? 
3. What genres make the most money?
4. Which producers generally do well? 
5. Is there a bias with producers in terms of genres that they do - Do specific producer only produce specific genre movies?
6. Will a longer run time bring higher profits?
7. Does the presence of a homepage indicate higher profits
8. Which languages rakes in high profit
9. Over a period of time, how have the profitability changed - Over the period of time, which genre of movies have grossed well ? 
10. If the movie speaks multiple languages, is the profit % higher?
11. On which day (Sun/Mon/Tue), if i release a movie, I will get high profits ? 
12. Which release months brings high profits ? 
13. Which is the most profitable and least collection of movies ? 